golang/cosmos/x/swingset/keeper/snapshotter.go

```diff
@@ -0,0 +1,421 @@
+	// completes while we check its status
+	activeSnapshot := snapshotter.activeSnapshot
+
+	if activeSnapshot == nil || activeSnapshot.startedResult == nil {
+		return nil
+	}
+
+	// Indicate to future callers that we already waited for the snapshot
+	// operation to have started so they can return immediately.
+	startedResult := activeSnapshot.startedResult
+	activeSnapshot.startedResult = nil
+
+	return <-startedResult
+}
+
+// SnapshotName returns the name of snapshotter, it should be unique in the manager.
+func (snapshotter *SwingsetSnapshotter) SnapshotName() string {
+	return types.ModuleName
+}
+
+// SnapshotFormat returns the default format the extension snapshotter use to encode the
+// payloads when taking a snapshot.
+// It's defined within the extension, different from the global format for the whole state-sync snapshot.
+func (snapshotter *SwingsetSnapshotter) SnapshotFormat() uint32 {
+	return SnapshotFormat
+}
+
+// SupportedFormats returns a list of formats it can restore from.
+func (snapshotter *SwingsetSnapshotter) SupportedFormats() []uint32 {
+	return []uint32{SnapshotFormat}
+}
+
+// SnapshotExtension writes extension payloads into the underlying protobuf stream.
+// This operation is invoked by the snapshot manager in the goroutine started by
+// `InitiateSnapshot`.
+func (snapshotter *SwingsetSnapshotter) SnapshotExtension(height uint64, payloadWriter snapshots.ExtensionPayloadWriter) (err error) {
+	defer func() {
+		// Since the cosmos layers do a poor job of reporting errors, do our own reporting
+		// `err` will be set correctly regardless if it was explicitly assigned or
+		// a value was provided to a `return` statement.
+		// See https://go.dev/blog/defer-panic-and-recover for details
+		if err != nil {
+			snapshotter.logger.Error("swingset snapshot extension failed", "err", err)
+		}
+	}()
+
+	if snapshotter.activeSnapshot == nil {
+		return errors.New("no active swingset snapshot")
+	}
+
+	if snapshotter.activeSnapshot.height != int64(height) {
+		return fmt.Errorf("swingset snapshot requested for unexpected height %d (expected %d)", height, snapshotter.activeSnapshot.height)
+	}
+
+	action := &snapshotAction{
+		Type:        "COSMOS_SNAPSHOT",
+		BlockHeight: snapshotter.activeSnapshot.height,
+		Request:     "retrieve",
+	}
+	out, err := snapshotter.blockingSend(action)
+
+	if err != nil {
+		return err
+	}
+
+	var exportDir string
+	err = json.Unmarshal([]byte(out), &exportDir)
+	if err != nil {
+		return err
+	}
+
+	defer os.RemoveAll(exportDir)
+
+	rawManifest, err := os.ReadFile(filepath.Join(exportDir, ExportManifestFilename))
+	if err != nil {
+		return err
+	}
+
+	var manifest exportManifest
+	err = json.Unmarshal(rawManifest, &manifest)
+	if err != nil {
+		return err
+	}
+
+	if manifest.BlockHeight != height {
+		return fmt.Errorf("snapshot manifest blockHeight (%d) doesn't match (%d)", manifest.BlockHeight, height)
+	}
+
+	fileToPayload := func(name string, filename string) error {
```

Comment:
Suggested rename and parameter refactoring:
```suggestion
	writeFileToPayload := func(fileName string, artifactName string) error {
```